cpp/tests/testnnevalcanary.cpp

```diff
@@ -276,7 +276,79 @@ struct GpuErrorStats {
   }
 };
 
-bool Tests::runFP16Test(NNEvaluator* nnEval, NNEvaluator* nnEval32, Logger& logger, int boardSize, int maxBatchSizeCap, bool verbose, bool quickTest, bool& fp32BatchSuccessBuf) {
+void saveBaseToFile(const std::vector<std::shared_ptr<NNOutput>>& base, const string& baseFileName, Logger& logger, bool verbose) {
+  assert(baseFileName != "");
+  std::ofstream outFile(baseFileName, std::ios::binary);
+
+  if (!outFile)
+    throw StringError("Unable to save base to: " + baseFileName);
+
+  size_t size = base.size();
+  outFile.write(reinterpret_cast<const char*>(&size), sizeof(size));
+
+  for (const auto& nnOutputPtr : base) {
+    if (nnOutputPtr) {
+      outFile.write(reinterpret_cast<const char*>(&nnOutputPtr->nnHash), sizeof(nnOutputPtr->nnHash));
```

Comment:
It's generally a bad idea to reinterpret_cast to const char* to do serialization of data.  It's not easy to verify from here that the data you're outputting are all primitives, and you have to worry about endianness differences between CPU platforms.  If the values are not primatives but structs, then you have to ensure that padding is consistent between all platforms, and that's very much not likely to always be true even on the same architecture (it can vary based on even compiler optimizations).

Reply:
I appreciate your concerns regarding the use of reinterpret_cast for serialization. However, this implementation is specifically designed for environments where both the architecture and compiler settings remain consistent. Users are responsible for ensuring the output data is saved and loaded in the same environment.

Reply:
I'm replacing this with json-based serialization. Generates larger files due to the output being in text format instead of binary, but it has the advantage of being cross-platform so that users can compare to a reference provided by someone else, as long as they're using the same neural net. Also the output is human readable if for some reason we need to debug the actual outputs, and the code in C++ for doing the json serialization is not very complicated.